csrc/optimization/move_split_cat.cpp

```diff
@@ -30,52 +115,76 @@ namespace {
 //
 // findCancelingSplit(out) returns `in`.
 //
-// Currently, the cat has to immediately follow the split. But we will soon
-// handle patterns like `split ->
-// a_chain_of_unary_ops_that_do_not_resize_the_split_dimension -> cat`.
-TensorView* findCancelingSplit(CatOp* cat) {
-  // nvFuser implements cat as PadOp->CatOp. We first locate these `PadOp`s.
-  std::vector<PadOp*> pads;
-  pads.reserve(cat->inputs().size());
-  for (Val* in : cat->inputs()) {
-    pads.push_back(in->definition()->as<PadOp>());
-  }
+// `cat` doesn't have to immediately follow the split. For example, when
+//
+//   s0 = in[:, :5]
+//   s1 = in[:, 5:]
+//   t0 = permute(s0)
+//   t1 = permute(s1)
+//   out = cat([t0, t1], dim=0)
+//
+// In addition to returning `in`, findCancelingSplit(out) puts `t0`'s defining
+// `permute` into `use_def_chain` so the caller can reconstruct `out` by
+// replaying `use_def_chain` (in reverse order) on `in`.
+TensorView* findCancelingSplit(CatOp* cat, std::vector<Expr*>& use_def_chain) {
+  NVF_CHECK(!cat->inputs().empty(), "`cat` has zero inputs: ", cat);
+
+  auto [pads, cat_axis] = getCatInputsAndAxis(cat);
 
-  // We then locate the `SliceOp`s right before these `PadOp`s.
-  std::vector<SliceOp*> slices;
-  slices.reserve(pads.size());
+  // `frontier` initially contains the preceding Exprs of `pads`. Then, we
+  // repeatedly try to move the frontier up in lockstep as long as Exprs in the
+  // frontier can be horizontally merged and applied on the unsplit tensor.
+  std::vector<Expr*> frontier;
+  frontier.reserve(pads.size());
   for (PadOp* pad : pads) {
-    auto* slice = dynamic_cast<SliceOp*>(pad->in()->definition());
-    if (slice == nullptr) {
+    frontier.push_back(pad->in()->definition());
+  }
+
+  // Exit the loop when any Expr in `frontier` is a slice or a null.
+  int64_t split_axis = cat_axis;
+  while (std::none_of(frontier.begin(), frontier.end(), [](Expr* e) {
+    return e == nullptr || e->isA<SliceOp>();
+  })) {
+    if (!horizontallyMergeable(frontier, std::ref(split_axis))) {
       return nullptr;
     }
-    slices.push_back(slice);
+    use_def_chain.push_back(frontier[0]);
+
+    // Advance the frontier in lockstep.
+    for (Expr*& e : frontier) {
+      NVF_ERROR(
+          e->inputs().size() == 1,
```

Comment:
ha, nvm. I see that we are only handling unary ops.

Reply:
Wouldn't squeeze/unsqueeze/broadcast still falls in here? I think we'll need to compensate split_axis for those as well... Otherwise, we might falsely cancel out split/concat on different axis.

Reply:
`horizontallyMergeable` returns false for anything that's not a set, so this optimization bails out when seeing squeeze/unsqueeze/broadcast between the split and the cat. Is that still a problem?